### PR TITLE
Add getBytes()

### LINKED ISSUE #76: FR: Alternative to getDownloadURL that doesn't expose a public download URL
### [REQUIRED] Describe your environment

  * Operating System version: [all]
  * Firebase SDK version: 4.1.3
  * Firebase Product: storage

### [REQUIRED] Describe the problem

Currently, the only way to get an image is to call [`getDownloadUrl`](https://firebase.google.com/docs/reference/js/firebase.storage.Reference#getDownloadURL). The download URL is unguessable but not affected by security rules, which means that once someone has the download URL, they could theoretically use a bot to create a huge number of downloads and run up a big Cloud Storage bill or hit the project quota.

 The Android SDK provides [`getFile`](https://firebase.google.com/docs/reference/android/com/google/firebase/storage/StorageReference.html#getFile(java.io.File)), which downloads the file directly to the device. Each download involves a security rules check ([source](https://stackoverflow.com/a/39920698/4816918)).

### Request

Provide a function like `getFile` that checks security rules every time.

Alternatively, perhaps a function that returns a [signed URL](https://cloud.google.com/storage/docs/access-control/signed-urls) (for example, `getSignedUrl(expireTime)`) could accomplish a similar goal.

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
Adds getBytes, getBlob and getStream APIs which allows file downloads without a public download URL. These bytes can then be used directly as an image source.

This PR contains a bunch of refactors to make this possible:

- There are now three types of Connections. `Connection<string>`, `Connection<ArrayBuffer>` and `Connection<ReadableStream>`.
- Each connection type has a distinct implementation for Node and Browsers.

Fixes #76

### CODE REVIEW

File: packages/storage/src/api.ts [hunk 1]
```diff
@@ -76,6 +77,28 @@ export {
...
 
+/**
+ * Downloads the data at the object's location. Returns an error if the object
+ * is not found.
+ *
+ * To use this functionality, you have to whitelist your app's origin in your
+ * Cloud Storage bucket. See also
+ * https://cloud.google.com/storage/docs/configuring-cors
+ *
+ * @public
```
Comment (Feiyang1) [inline packages/storage/src/api.ts:88]: @egilmorez can you please scan the PR and review the comments with the `@public` tag? Thanks!

File: packages/storage/src/implementation/request.ts [hunk 2]
```diff
@@ -46,15 +40,24 @@ export interface Request<T> {
   cancel(appDelete?: boolean): void;
 }
 
-class NetworkRequest<T> implements Request<T> {
-  private pendingConnection_: Connection | null = null;
+/**
+ * Handles network logic for all Storage Requests, including error reporting and
+ * retries with backoff.
+ *
+ * @param I - the type of the backend's network response (always `string`,
```
Comment (Feiyang1) [inline packages/storage/src/implementation/request.ts:47]: Sounds like we should add a constraint to the generic parameter? e.g. `<I extends string | ArrayBuffer |  ReadableStream, O>`?

File: packages/storage/src/platform/browser/connection.ts [hunk 3]
```diff
@@ -126,12 +131,96 @@ export class XhrConnection implements Connection {
...
+export function newTextConnection(): Connection<string> {
+  return textFactoryOverride ? textFactoryOverride() : new XhrTextConnection();
+}
+
+export class XhrBytesConnection extends XhrConnection<ArrayBuffer> {
+  private data_?: ArrayBuffer;
+
+  initXhr(): void {
+    // We use Blob here instead of ArrayBuffer to ensure that this code works
+    // in Opera.
```
Comment (Feiyang1) [inline packages/storage/src/platform/browser/connection.ts:156]: Nice! Do you have a link to the limitation in Opera?

File: packages/storage/src/platform/connection.ts [hunk 4]
```diff
@@ -16,16 +16,36 @@
...
 
-export function injectTestConnection(factory: (() => Connection) | null): void {
+export function newTextConnection(): Connection<string> {
   // This file is only used under ts-node.
-  nodeInjectTestConnection(factory);
+  return nodeNewTextConnection();
+}
+
+export function newBytesConnection(): Connection<ArrayBuffer> {
+  // This file is only used under ts-node.
```
Comment (Feiyang1) [inline packages/storage/src/platform/connection.ts:39]:
```suggestion
  // This file is only used in Node.js tests using ts-node.
```

File: packages/storage/src/platform/connection.ts [hunk 5]
```diff
@@ -16,16 +16,36 @@
...
+  return nodeNewBytesConnection();
+}
+
+export function newBlobConnection(): Connection<Blob> {
+  // This file is only used under ts-node.
```
Comment (Feiyang1) [inline packages/storage/src/platform/connection.ts:44]:
```suggestion
  // This file is only used in Node.js tests using ts-node.
```

File: packages/storage/src/platform/connection.ts [hunk 6]
```diff
@@ -16,16 +16,36 @@
...
+  return nodeNewBlobConnection();
+}
+
+export function newStreamConnection(): Connection<NodeJS.ReadableStream> {
+  // This file is only used under ts-node.
```
Comment (Feiyang1) [inline packages/storage/src/platform/connection.ts:49]:
```suggestion
  // This file is only used in Node.js tests using ts-node.
```

File: packages/storage/src/platform/node/connection.ts [hunk 7]
```diff
@@ -120,12 +115,88 @@ export class FetchConnection implements Connection {
...
+    body?: ArrayBufferView | Blob | string,
+    headers?: Record<string, string>
+  ): Promise<void> {
+    if (this.sent_) {
+      throw internalError('cannot .send() more than once');
+    }
+    this.sent_ = true;
+
+    try {
+      const response = await nodeFetch(url, {
```
Comment (Feiyang1) [inline packages/storage/src/platform/node/connection.ts:161]:
```suggestion
      const response = await this.fetch_(url, {
```

so you can stub fetch easily in tests.

File: .changeset/clever-eggs-relate.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+---
+"@firebase/storage": minor
```
Comment (Feiyang1) [inline .changeset/clever-eggs-relate.md:2]:
```suggestion
"@firebase/storage": minor
"firebase": minor
```

File: packages/firestore/test/unit/specs/query_spec.test.ts [hunk 9]
```diff
@@ -100,6 +100,27 @@ describeSpec('Queries:', [], () => {
       });
   });
 
+  Caused by: java.lang.AssertionError: expected:<
```
Comment (Feiyang1) [inline packages/firestore/test/unit/specs/query_spec.test.ts:103]: what is it?

Comment (schmidt-sebastian) [reply packages/firestore/test/unit/specs/query_spec.test.ts]: This .... ehmm... seems unrelated. Removed.

**Review by schmidt-sebastian — COMMENTED** [review]
I cleaned up the PR and simplified some of the logic. It should now be ready for another round.

**Review by Feiyang1 — APPROVED** [review]

File: packages/storage/src/api.ts (hunk 1 above)
Comment (egilmorez) [reply packages/storage/src/api.ts:88]: I was hoping @markarndt could give that a go . . . but if he's too busy and there's urgency, I can take a look this afternoon for sure.

File: packages/storage/src/api.ts [hunk 10]
```diff
@@ -76,6 +77,28 @@ export {
...
+ * @param ref - StorageReference where data should be download.
```
Comment (markarndt) [inline packages/storage/src/api.ts:89]: "..should be downloaded."

**Review by markarndt — APPROVED** [review]

**Pull request merged**